src/merge_events.cpp

```diff
@@ -0,0 +1,205 @@
+    heap_[i] = event(l)<event(r)? heap_[l]: heap_[r];
+}
+
+// The tree is stored using the standard heap indexing scheme.
+
+unsigned tourney_tree::parent(unsigned i) const {
+    return (i-1)>>1;
+}
+unsigned tourney_tree::left(unsigned i) const {
+    return (i<<1) + 1;
+}
+unsigned tourney_tree::right(unsigned i) const {
+    return left(i)+1;
+}
+unsigned tourney_tree::leaf(unsigned i) const {
+    return i+leaves_-1;
+}
+bool tourney_tree::is_leaf(unsigned i) const {
+    return i>=leaves_-1;
+}
+const unsigned& tourney_tree::id(unsigned i) const {
+    return heap_[i].first;
+}
+postsynaptic_spike_event& tourney_tree::event(unsigned i) {
+    return heap_[i].second;
+}
+const postsynaptic_spike_event& tourney_tree::event(unsigned i) const {
+    return heap_[i].second;
+}
+
+unsigned tourney_tree::next_power_2(unsigned x) const {
+    unsigned n = 1;
+    while (n<x) n<<=1;
+    return n;
+}
+
+} // namespace impl
+
+void merge_events(time_type t0, time_type t1,
+                  const pse_vector& lc, pse_vector& events,
+                  std::vector<event_generator_ptr>& generators,
+                  pse_vector& lf)
+{
+    using std::distance;
+    using std::lower_bound;
+
+    // Sort events from the communicator in place.
+    util::sort(events);
+
+    // Clear lf to store merged list.
+    lf.clear();
+
+    // Merge the incoming event sequences into a single vector in lf
+    if (generators.size()) {
+        // Handle the case where the cell has event generators.
+        // This is performed in two steps:
+        //  Step 1 : Use tournament tree to merge all events in lc, events and
+        //           generators to be delivered in the time interval [t₀, t₁).
+        //  Step 2 : Use std::merge to append events in lc and events with
+        //           delivery times in the interval [t₁, ∞).
+        EXPECTS(generators.size()>2u);
+
+        // Make an event generator with all the events in events.
+        generators[0] = make_event_generator<seq_generator<pse_vector>>(events);
+
+        // Make an event generator with all the events in lc with time >= t0
+        auto lc_it = lower_bound(lc.begin(), lc.end(), t0, event_time_less());
+        auto lc_range = util::make_range(lc_it, lc.end());
+        generators[1] = make_event_generator<seq_generator<decltype(lc_range)>>(lc_range);
+
+        // Perform k-way merge of all events in events, lc and the generators
+        // that are due to be delivered in the interval [t₀, t₁)
+        impl::tourney_tree tree(generators);
+        while (!tree.empty(t1)) {
+            lf.push_back(tree.head());
+            tree.pop();
+        }
+
+        // Find first event in lc with delivery time >= t1
+        lc_it = lower_bound(lc.begin(), lc.end(), t1, event_time_less());
+        // Find first event in events with delivery time >= t1
+        auto ev_it = lower_bound(events.begin(), events.end(), t1, event_time_less());
+        const auto m = lf.size();
```

Comment:
It is the number of events already in `lf`.
I kept the name short because:
* it is used just on the very next line
* I use a seperate variable because the code on the next line was becoming to verbose.